src/main/java/com/google/cloud/spanner/pgadapter/utils/MutationBuilder.java

```diff
@@ -0,0 +1,208 @@
+                  + ":"
+                  + "\""
+                  + recordValue
+                  + "\"");
+        } catch (IllegalArgumentException e) {
+          writeToSpanner(connectionHandler);
+          createErrorFile(payload);
+          throw new SQLException("Invalid input syntax for column \"" + columnName + "\"");
+        } catch (Exception e) {
+          writeToSpanner(connectionHandler);
+          createErrorFile(payload);
+          throw e;
+        }
+      }
+      this.mutations.add(builder.build()); // Add write builder to mutation list
+      this.mutationCount += record.size(); // Increment the number of mutations being added
+      this.rowCount++; // Increment the number of COPY rows by one
+    }
+    this.batchSize += payload.length; // Increment the batch size based on payload length
+  }
+
+  /**
+   * @return True if adding payload to current batch will fit under mutation limit and batch size
+   *     limit, false otherwise.
+   */
+  private boolean payloadFitsInCurrentBatch(int rowMutationCount, int payloadLength) {
+    return (this.mutationCount + rowMutationCount <= MUTATION_LIMIT
+        && this.batchSize + payloadLength <= BATCH_LIMIT);
+  }
+
+  /** @return list of CSVRecord rows parsed with CSVParser from CopyData payload byte array */
+  private List<CSVRecord> parsePayloadData(byte[] payload) throws IOException {
+    String copyData = new String(payload, StandardCharsets.UTF_8).trim();
+    CSVParser parser;
+    if (this.hasHeader && !this.isHeaderParsed) {
+      parser = CSVParser.parse(copyData, this.format.withFirstRecordAsHeader());
+      this.isHeaderParsed = true;
+    } else {
+      parser = CSVParser.parse(copyData, this.format);
+    }
+    return parser.getRecords();
+  }
+
+  /**
+   * Write list of mutations to Spanner and reset for next batch
+   *
+   * @return count of the number of rows updated.
+   */
+  public int writeToSpanner(ConnectionHandler connectionHandler) throws SQLException {
+    if (this.mutations.isEmpty()) {
+      return 0;
+    }
+    Connection connection = connectionHandler.getJdbcConnection();
+    CloudSpannerJdbcConnection spannerConnection =
+        connection.unwrap(CloudSpannerJdbcConnection.class);
+    spannerConnection.write(this.mutations); // Write mutation to spanner
+    // Reset mutations, mutation counter, and batch size count for a new batch
+    this.mutations = new ArrayList<>();
+    this.mutationCount = 0;
+    this.batchSize = 0;
+    return this.rowCount;
+  }
+
+  public void createErrorFile(byte[] payload) throws IOException {
+    File unsuccessfulCopy = new File("./output.txt");
+    if (unsuccessfulCopy.createNewFile()) {
+      this.fileWriter = new FileWriter("output.txt");
+      writeToErrorFile(payload);
+    } else {
+      System.err.println("File " + unsuccessfulCopy.getName() + " already exists");
+    }
+  }
+
+  public void writeToErrorFile(byte[] payload) throws IOException {
```

Comment:
public?

Reply:
This will be accessed by the Copy messages (CopyDone, CopyData, ... not included in this PR).